### PR TITLE
Allow to flush segment in separate thread

### PR DESCRIPTION
**Pull request by IvanPleshkov**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/segment.rs [hunk 1]
```diff
@@ -574,45 +591,61 @@ impl SegmentEntry for Segment {
...
+        let persisted_version = self.persisted_version.clone();
+        let flush_op = move || {
+            // Flush mapping first to prevent having orphan internal ids.
+            id_tracker.borrow().flush_mapping().map_err(|err| {
+                OperationError::service_error(&format!(
+                    "Failed to flush id_tracker mapping: {}",
+                    err
+                ))
+            })?;
+            vector_storage.borrow().flush().map_err(|err| {
```
Comment (generall) [inline lib/segment/src/segment.rs:616]: will panic on concurrent borrow_mut

**Review by generall — CHANGES_REQUESTED** [review]

File: lib/segment/src/segment.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/segment/src/segment.rs]: Fixed

File: lib/segment/src/segment.rs [hunk 2]
```diff
@@ -248,6 +251,20 @@ impl Segment {
...
+    // Joins flush thread if exists
+    // Returns lock to guarantee that there will no any fush in different thread
+    fn lock_flushing(
+        &self,
+    ) -> parking_lot::MutexGuard<Option<JoinHandle<OperationResult<SeqNumberType>>>> {
+        let mut lock = self.flush_thread.lock();
+        let mut join_handle: Option<JoinHandle<OperationResult<SeqNumberType>>> = None;
+        std::mem::swap(&mut join_handle, &mut lock);
+        if let Some(join_handle) = join_handle {
+            join_handle.join().unwrap().unwrap();
```
[the remark this answers is not in the document]
Comment (IvanPleshkov) [reply lib/segment/src/segment.rs]: Fixed by @generall

**Comment by generall** [comment]
@IvanPleshkov can you please share benchmarks result here?

**Comment by IvanPleshkov** [comment]
> @IvanPleshkov can you please share benchmarks result here?

Test case: go grpc client, upsert 100.000 points without payload, dim=1000, distance=euclid. Flush interval = 5sec. I take the worst upserts.

Worst upserts in master:
```
2022/08/11 09:23:16 25.992231ms
2022/08/11 09:23:16 27.316313ms
2022/08/11 09:23:16 27.66545ms
2022/08/11 09:23:16 29.298539ms
2022/08/11 09:23:16 183.192019ms
2022/08/11 09:23:16 236.495462ms
2022/08/11 09:23:16 275.627032ms
2022/08/11 09:23:16 280.808815ms
2022/08/11 09:23:16 606.951983ms
```

Worst upserts in branch
```
2022/08/11 09:17:07 18.332058ms
2022/08/11 09:17:07 18.591699ms
2022/08/11 09:17:07 20.369265ms
2022/08/11 09:17:07 20.618688ms
2022/08/11 09:17:07 21.697782ms
2022/08/11 09:17:07 24.675179ms
2022/08/11 09:17:07 28.206836ms
2022/08/11 09:17:07 30.796368ms
```

**Comment by IvanPleshkov** [comment]
Also I changed flush interval to 5sec instead of 1sec

File: lib/segment/src/segment.rs [hunk 3]
```diff
@@ -574,45 +604,64 @@ impl SegmentEntry for Segment {
...
-            OperationError::service_error(&format!("Failed to flush database: {}", err))
-        })?;
-
-        *persisted_version = state.version;
-
-        Ok(state.version)
+        if sync {
+            flush_op()
+        } else {
+            *background_flush_lock = Some(std::thread::spawn(flush_op));
```
Comment (agourlay) [inline lib/segment/src/segment.rs:662]: Minor suggestion:

I think it is a good practice to name our threads to help us identify issues later.

```suggestion
            *background_flush_lock = Some(
              std::thread::Builder::new()
                .name("background_flush".to_string())
                .spawn(flush_op)
                .unwrap());
```

**Comment by agourlay** [comment]
Are we also using `flush(false)` somewhere in the unit tests for sanity check?

**Comment by agourlay** [comment]
sorry, my suggestion was not properly formatted :grimacing:

**Comment by IvanPleshkov** [comment]
> sorry, my suggestion was not properly formatted 😬

fixed

**Comment by IvanPleshkov** [comment]
> Are we also using `flush(false)` somewhere in the unit tests for sanity check?

That's an untrivial task because we don't have any notification about finishing background flush. What do you think @generall ?

**Comment by generall** [comment]
> Are we also using flush(false) somewhere in the unit tests for sanity check?

it is used as a part of optimizer. But I think writing an explicit test should not be too complicated

**Comment by agourlay** [comment]
Once I merge https://github.com/qdrant/qdrant/pull/942 you will be able to rebase and remove the remaining Clippy failures.

**Review by generall — APPROVED** [review]

**Pull request merged**